### ISSUE TITLE
Document how to generate heap dumps for a running JVM

### ISSUE THREAD

**Issue by avermeer** [eclipse/openj9-docs#309]
**Subject:**
Current OpenJ9 documentation gives many information related to heap dumps with -Xdump option, but noting to generate an heap dump file like the following:

`jcmd `cat /var/run/tomee.pid` GC.heap_dump /mnt/tomee`date +"%Y_%m_%d_%H_%M"`.dmp`

A team mate has gone as far as adding a REST endpoint in his application using internal classes to generate such done (in ZIP format because unlike with HotSpot, these dumps are really huge):

```
      public Response appServerHeapDump(@Context HttpServletRequest httpRequest) {
            if (!isLocal(httpRequest))
                  return Response.status(Status.NOT_FOUND).build();
            JsonObjectBuilder json = Json.createObjectBuilder();
            try {
                  File coreFile = new File((String) Class.forName("com.ibm.jvm.Dump").getMethod("systemDumpToFile").invoke(null));
                  File outFile = new File(System.getProperty("java.io.tmpdir", "/tmp"), coreFile.getName() + ".gz");
                  try (InputStream in = new BufferedInputStream(new FileInputStream(coreFile));
                              OutputStream out = new GZIPOutputStream(new BufferedOutputStream(new FileOutputStream(outFile)))) {
                        in.transferTo(out);
                        json.add("file", String.valueOf(outFile));
                  } catch (Exception e1) {
                        json.add("error", String.valueOf(e1));
                  } finally {
                        coreFile.delete();
                  }
            } catch (Exception e) {
                  json.add("error", String.valueOf(e));
            }
            return Response.ok().entity(json.build()).build();
      }
```

**Suggestion:**

It should be easy for people using OpenJ9 to understand how to generate a heap dump from a running process, so that this heap dump can be analyzed later using tools such as Eclipse MAT.

It looks like kill -3 <process ID of the running JVM> does not generates such dump.
I am aware of https://github.com/eclipse/openj9/issues/5164 but in the meantime, we need documentation to explain how to generate this dump with  current OpenJ9 versions.

Kind regards,
Alexandre

**Comment by pshipton** [comment]
A few notes:
- `-Xdump:what` shows the default settings
- `kill -QUIT <pid>` will create a heapdump if -Xdump has been configured to do so, but this is not the default behavior
- `-Xdump:heap:events=user,request=exclusive+compact+prepwalk` will configure -Xdump to create a heapdump on the user signal (`kill -QUIT <pid>`)

**Comment by avermeer** [comment]
@pshipton  thank you very much for your proposal to use `-Xdump:heap:events=user,request=exclusive+compact+prepwalk`, to make `kill -QUIT <pid>` generate an heap dump just like HotSpot does: I tried it, works like charm.

I feel that this setup should be documented in the HotSpot transition documentation.

Would you agree?

Kind regards,
Alexandre Vermeerbergen

**Comment by avermeer** [comment]
Hello Again,

Sorry I spoke too fast in my previous post @pshipton : the option you proposed did not generate an heap dump compatible with Eclipse MAT, here's the error I got from Eclipse MAT 1.8.1 with a dump generated from an application server process:

```
Error opening heap dump 'tomcat20190628-152255.dmp'. Check the error log for further details.
Error opening heap dump 'tomcat20190628-152255.dmp'. Check the error log for further details.
Not a HPROF heap dump (java.io.IOException)
Not a HPROF heap dump
```

I find strange that Eclipse MAT is not able to read OpenJ9's heap dumps (both are Eclipse projects).
Am I missing something to be able to use Eclipse MAT to analyze OpenJ9 dumps?

Kind regards,
Alexandre Vermeerbergen

**Comment by pshipton** [comment]
MAT has a plugin that enables it to read OpenJ9 dumps. See https://www.eclipse.org/openj9/docs/dump_heapdump/

If that doesn't work, try adding `opts=CLASSIC` to the -Xdump option
`-Xdump:heap:events=user,request=exclusive+compact+prepwalk,opts=CLASSIC`
or
`-Xdump:heap:events=user,request=exclusive+compact+prepwalk,opts=PHD+CLASSIC`

**Comment by avermeer** [comment]
Hello @pshipton 

Here's what I tested:
* `-Xdump:heap:events=user,request=exclusive+compact+prepwalk,opts=CLASSIC,file=/tmp/myapp.dmp`
  => Trying to read this dump using Eclipse MAT without the DTFJ plugin says:
```
   Error opening heap dump 'tomcat20190629-061127.classic'. Check the error log for further details.
Error opening heap dump 'tomcat20190629-061127.classic'. Check the error log for further details.
Invalid HPROF file header. (java.io.IOException)
Invalid HPROF file header.
```

*  Xdump:heap:events=user,request=exclusive+compact+prepwalk,opts=PHD+CLASSIC,file=/tmp
/myapp.dmp.
  With this test, it's strange because at first ,file /tmp/myapp.dmp contains a text file, then few seconds after ,file /tmp/myapp.dmp content is replaced by a binary file
   => Trying to read this dump using Eclipse MAT without the DTFJ plugin says:
```
Error opening heap dump 'tomcat20190629-131938.dmp'. Check the error log for further details.
Error opening heap dump 'tomcat20190629-131938.dmp'. Check the error log for further details.
Not a HPROF heap dump (java.io.IOException)
Not a HPROF heap dump
```

* Using Eclipse MAT with the DTFJ plugin, all dumps can be read

My thoughts:
1. It seems that OpenJ9 cannot generate a dump in HPROF format, is it correct?
2. Given that Eclipse MAT is part of Eclipse projects, just like OpenJ9 is an Eclipse project, wouldn't it be a better user experience to have Eclipse MAT being able to read OpenJ9 dumps without requiring an external plug-in?
3. The behavior of PHD+CLASSIC sounds very strange to me: since the initial text file is replaced by the PHD dump in the output file, what is the difference between PHD and PHD+CLASSIC?

Kind regards,
Alexandre Vermeerbergen

**Comment by pshipton** [comment]
I thought maybe MAT out of the box could read the CLASSIC format, but apparently not.

There is some information about using the hprof agent here https://www.ibm.com/support/knowledgecenter/SSYKE2_8.0.0/com.ibm.java.80.doc/diag/tools/hprof.html but OpenJ9 doesn't have built in support to generate heap dumps in the HPROF format. There may be licensing problems related to doing that. Note the hprof agent was removed in Java 9 (https://openjdk.java.net/jeps/240), so only Java 8 contains this agent.

Although MAT and OpenJ9 are Eclipse projects, currently the DTFJ agent plugin is supplied by IBM and is not an Eclipse project. Agreed it would be better to have it contributed to Eclipse and included in MAT by default. I hope some day we can make this happen.

I expected PHD+CLASSIC to create two different files, one in PHD format and one in CLASSIC (txt) format. This does occur when the `file=` option is not used, i.e. both `heapdump.20190703.073301.24368.0001.txt` and `heapdump.20190703.073301.24368.0001.phd` are created. It seems using PHD+CLASSIC is incompatible with the `file=` option.

**Comment by SueChaplain** [comment]
@pshipton - in your opinion, is there anything further we could add to the docs to clear up any confusion here. If not, I propose to close this issue.

**Comment by pshipton** [comment]
jcmd is implemented on OpenJ9 from the 0.16 release for Java 13 which is already released, and the 0.17 release for Java 8, 11+, due for release ~Oct 18.

**Comment by deleze** [comment]
As @avermeer say, using the IBM plugin works.
To analyze Java 8 OpenJ9 dump with Eclipse Memory Analyser, install the IBM plugin from menu "Help > Install new software…" and search for "IBM"
<img width="982" alt="image" src="https://github.com/eclipse-openj9/openj9-docs/assets/2012495/f3eb5da9-2301-4fd8-8907-964dcb5247f1">
Then restart the Eclipse Memory Analyser and open the dump.
